Mathlib/GameTheory/Auction/Secondprice.lean

```diff
@@ -0,0 +1,271 @@
+    apply Finset.le_sup'
+    simp only [Finset.mem_univ, Finset.mem_erase, and_true]
+    exact (Ne.symm H)
+  exact Real.partialOrder.proof_4 (B b i) (Finset.sup' Finset.univ maxb.proof_1 b) H1 H2
+
+/-- Utility is non-negative if the bid equals the valuation. -/
+lemma utility_nneg (i : a.I) (H : b i = a.v i) : 0 ≤ utility b i := by
+  by_cases H2: i = winner b
+  · rw [utility]
+    simp only [H2]
+    rw [← H2, ← H, H2]
+    rw [winner_take_max b]
+    apply sub_nonneg.mpr
+    rw [secondprice]
+    apply Finset.sup'_le
+    simp only [Finset.mem_univ, Finset.mem_erase, and_true]
+    intro j _
+    rw [maxb]
+    simp only [Finset.le_sup'_iff, Finset.mem_univ, true_and]
+    use j
+  · rw[utility, if_neg H2]
+
+/-- Proves that the strategy of bidding one's valuation is a dominant strategy for `i`. -/
+theorem valuation_is_dominant (i : a.I ) : dominant i (a.v i) := by
+  intro b b' hb hb'
+  by_cases H : i = winner b'
+  · by_cases H1 : B b' i < a.v i
+    · have h_winner_b : i = winner b := gt_wins b i (λ j hj => by
+      rw [hb, hb']
+      have HBi: b' j ≤ B b' i := by
+        rw [B]
+        simp only [Finset.mem_univ, not_true, ge_iff_le, Finset.le_sup'_iff,
+                  Finset.mem_erase, ne_eq, and_true]
+        use j
+        simp only [le_refl, and_true]
+        rw [← ne_eq,ne_comm]
+        exact hj
+      exact gt_of_gt_of_ge H1 HBi
+      exact id (Ne.symm hj))
+      rw [utility_winner _ h_winner_b, utility_winner _ H]
+      have h_secondprice_eq : secondprice b = secondprice b' := by
+        repeat rw [secondprice]
+        rw[← h_winner_b, ← H]
+        repeat rw [B]
+        apply Finset.sup'_congr (delete_i_nonempty i) (rfl)
+        intro j hj
+        rw [Finset.mem_erase] at hj
+        exact hb' j hj.1
+      · rw [h_secondprice_eq]
+    · rw [utility,← H]
+      simp only [ite_true, ge_iff_le, tsub_le_iff_right]
+      simp only [gt_iff_lt, not_lt] at H1
+      rw [secondprice, ← H]
+      have := utility_nneg b i hb
+      exact le_add_of_nonneg_of_le this H1
+  · have := utility_nneg b i hb
+    convert this
+    simp [utility, H]
+
+/-- Computes the utility for a first price auction where the winner pays their bid. -/
+noncomputable def Utility.FirstPrice (i: a.I) : ℝ := if i = winner b then a.v i - b i else 0
+
+/-- If `i` is the winner in a first price auction, utility is their valuation minus their bid. -/
+lemma utility_first_price_winner(i: a.I) (H: i = winner b): Utility.FirstPrice b i = a.v i-b i := by
+  rw [H]
+  simp only [Utility.FirstPrice, if_true]
+
+/-- If `i` is not the winner in a first price auction, their utility is 0. -/
+lemma utility_first_price_loser(i: a.I) (H: i ≠ winner b): Utility.FirstPrice b i = 0 := by
+  rw [Utility.FirstPrice]
+  simp only [H, if_false]
+
+
+/-- Defines a dominant strategy in the context of a first price auction. -/
+def Dominant.FirstPrice (i : a.I) (bi : ℝ) : Prop :=
+  ∀ b b': a.I → ℝ, (b i = bi) → (∀ j: a.I, j ≠ i → b j = b' j)
```

Comment:
This pair of conditions `b i = bi` and `∀ j, j ≠ i → b j = b' j` comes up a few times. Should it be given a name?